File: backend/services/positioningService.js
const AccessPoint = require('../models/accessPointModel');
const Room = require('../models/roomModel');

/**
 * Triangulate coordinates using the Weighted Centroid Algorithm over scanned access points.
 * @param {Array<Object>} scans - List of AP readings: [{ macAddress: String, rssi: Number }]
 * @returns {Object|null} { x, y, floor } computed coordinate context, or null on failure
 */
async function triangulateRSSI(scans) {
  if (!scans || scans.length === 0) return null;

  try {
    // Filter scan inputs to resolve registered APs in our database
    const macs = scans.map(s => s.macAddress);
    const aps = await AccessPoint.findAll({
      where: { macAddress: macs }
    });

    if (aps.length === 0) return null;

    let sumX = 0;
    let sumY = 0;
    let sumWeights = 0;
    let computedFloor = aps[0].floor; // Default to the floor of detected APs

    for (const scan of scans) {
      const ap = aps.find(a => a.macAddress === scan.macAddress);
      if (!ap) continue;

      // Filter out extremely weak signals (below -85 dBm) to prevent coordinate inflation
      if (scan.rssi < -85) continue;

      // Weight calculation: conversion of negative RSSI decibels to linear power weight
      // Closer APs (e.g. -50 dBm) yield significantly larger weights than further ones (e.g. -80 dBm)
      const weight = Math.pow(10, (scan.rssi + 100) / 10); // Standard HSL baseline shifts for positive weights

      sumX += ap.x * weight;
      sumY += ap.y * weight;
      sumWeights += weight;
    }

    if (sumWeights === 0) return null;

    const x = sumX / sumWeights;
    const y = sumY / sumWeights;

    return { x, y, floor: computedFloor };

  } catch (error) {
    console.error('[TRIANGULATION ERROR] Triangulation process failed:', error);
    return null;
  }
}

/**
 * Match coordinates to a registered room zone based on floor coordinates boundaries.
 * @param {Number} x - Computed X coordinate
 * @param {Number} y - Computed Y coordinate
 * @param {Number} floor - Computed floor level
 * @returns {Object|null} Room record on match, else null
 */
async function matchRoom(x, y, floor) {
  try {
    const matchedRoom = await Room.findOne({
      where: {
        floor: floor
      }
    });
    
    // In local SQLite setups, we can fetch all rooms on the floor and do coordinate checks
    const rooms = await Room.findAll({ where: { floor } });
    
    for (const room of rooms) {
      if (x >= room.xMin && x <= room.xMax && y >= room.yMin && y <= room.yMax) {
        return room;
      }
    }
    
    return null;
  } catch (error) {
    console.error('[ROOM MATCHER ERROR] Room matching process failed:', error);
    return null;
  }
}

module.exports = {
  triangulateRSSI,
  matchRoom
};

File: backend/src/main/java/com/evacsense/service/PositioningService.java
package com.evacsense.service;

import com.evacsense.model.AccessPoint;
import com.evacsense.model.Room;
import com.evacsense.repository.AccessPointRepository;
import com.evacsense.repository.RoomRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class PositioningService {

    @Autowired
    private AccessPointRepository accessPointRepository;

    @Autowired
    private RoomRepository roomRepository;

    public static class ScanInput {
        private String macAddress;
        private Integer rssi;

        public ScanInput() {}

        public ScanInput(String macAddress, Integer rssi) {
            this.macAddress = macAddress;
            this.rssi = rssi;
        }

        public String getMacAddress() { return macAddress; }
        public void setMacAddress(String macAddress) { this.macAddress = macAddress; }

        public Integer getRssi() { return rssi; }
        public void setRssi(Integer rssi) { this.rssi = rssi; }
    }

    public static class PositionResult {
        public float x;
        public float y;
        public int floor;

        public PositionResult(float x, float y, int floor) {
            this.x = x;
            this.y = y;
            this.floor = floor;
        }
    }

    public PositionResult triangulateRSSI(List<ScanInput> scans) {
        if (scans == null || scans.isEmpty()) return null;

        try {
            List<String> macs = scans.stream().map(ScanInput::getMacAddress).collect(Collectors.toList());
            List<AccessPoint> aps = accessPointRepository.findByMacAddressIn(macs);

            if (aps.isEmpty()) return null;

            double sumX = 0;
            double sumY = 0;
            double sumWeights = 0;
            int computedFloor = aps.get(0).getFloor(); // Default to the floor of first matched AP

            for (ScanInput scan : scans) {
                Optional<AccessPoint> matchedAp = aps.stream()
                        .filter(a -> a.getMacAddress().equalsIgnoreCase(scan.getMacAddress()))
                        .findFirst();

                if (matchedAp.isEmpty()) continue;

                // Filter out extremely weak signals (below -85 dBm) to prevent coordinate inflation
                if (scan.getRssi() < -85) continue;

                // Weight calculation: linear power weight
                double weight = Math.pow(10, (scan.getRssi() + 100) / 10.0);

                sumX += matchedAp.get().getX() * weight;
                sumY += matchedAp.get().getY() * weight;
                sumWeights += weight;
            }

            if (sumWeights == 0) return null;

            float x = (float) (sumX / sumWeights);
            float y = (float) (sumY / sumWeights);

            return new PositionResult(x, y, computedFloor);

        } catch (Exception e) {
            System.err.println("[TRIANGULATION ERROR] Triangulation process failed: " + e.getMessage());
            return null;
        }
    }

    public Room matchRoom(float x, float y, int floor) {
        try {
            List<Room> rooms = roomRepository.findByFloor(floor);
            
            for (Room room : rooms) {
                if (x >= room.getxMin() && x <= room.getxMax() && y >= room.getyMin() && y <= room.getyMax()) {
                    return room;
                }
            }
            return null;
        } catch (Exception e) {
            System.err.println("[ROOM MATCHER ERROR] Room matching process failed: " + e.getMessage());
            return null;
        }
    }
}
